Kernel 4.0 / Debian: make programs fails with "Undefined reference to `build_channel_array'"

According to README, since I am running newer kernel than 3.16 I should only type `make programs`? 
When doing so, I am getting linking error:

```bash
$ make programs
cd sensors && make all
make[1]: Entering directory '/home/me/yoga-laptop/sensors'
cc generic_buffer.c  -o generic_buffer
/tmp/ccpVnbqv.o: dans la fonction « main »:
generic_buffer.c:(.text+0x139a): référence indéfinie vers « build_channel_array »
collect2: error: ld returned 1 exit status
Makefile:13: recipe for target 'generic_buffer' failed
make[1]: *** [generic_buffer] Error 1
make[1]: Leaving directory '/home/me/yoga-laptop/sensors'
Makefile:9: recipe for target 'programs' failed
make: *** [programs] Error 2
```

-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Hmm.  I'm not seeing that.

build_channel_array is defined in iio_utils.h as a inline function, so there
should be no linking issue at all.

peter


On 05/17/2015 01:49 PM, Gautier Pelloux-Prayer wrote:
> According to README, since I am running newer kernel than 3.16 I should
> only type |make programs|? When doing so, I am getting linking error:
> 
> $ make programs cd sensors && make all make[1]: Entering directory
> '/home/me/yoga-laptop/sensors' cc generic_buffer.c  -o generic_buffer 
> /tmp/ccpVnbqv.o: dans la fonction « main »: 
> generic_buffer.c:(.text+0x139a): référence indéfinie vers «
> build_channel_array » collect2: error: ld returned 1 exit status 
> Makefile:13: recipe for target 'generic_buffer' failed make[1]: ***
> [generic_buffer] Error 1 make[1]: Leaving directory
> '/home/me/yoga-laptop/sensors' Makefile:9: recipe for target 'programs'
> failed make: *** [programs] Error 2
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/pfps/yoga-laptop/issues/39>.
> 
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQEcBAEBCAAGBQJVXUIvAAoJECjN6+QThfjzbOQIAL+jXL2D3ucQq2dgU1S5X9go
DKRwoI9ficW2T8PENw2Y0snZaxYy6zuRSUxCT3+yi4eLO4jz78bw0bA3N3ZjIsKj
5IogfdA8gqGHTmVkPAZwYik81BSZeX53Nze1QP+n1/VPrKDA7AqimZP7+MJtYhRT
sjIg/H2CrweomnGhvjLBS8zXvayF2OydSbPKLntz2zN6mAcV6OLL7mPy5CpSxiOL
5Kd3QAJYBNMw2Uo56NjAQuVVAz7ssM+uHTwsXQAZgN+GO8v53go+QUzfqyn4mYFn
3JxYCLmnXNOHV+QUHHvDlE6lTlbnTS7sg7R2MD9qL8ja/JKqZhoqPpll4JPBsj0=
=KNFO
-----END PGP SIGNATURE-----

--- a/sensors/generic_buffer.c
+++ b/sensors/generic_buffer.c
@@ -35,6 +35,27 @@
 #include <inttypes.h>
 #include <syslog.h> /*pfps*/
 #include "iio_utils.h"
+extern inline int build_channel_array(const char *device_dir,
+                  struct iio_channel_info **ci_array,
+                  int *counter);
+extern inline int iioutils_break_up_name(const char *full_name,
+                  char **generic_name);
+extern inline int iioutils_get_param_float(float *output,
+                    const char *param_name,
+                    const char *device_dir,
+                    const char *name,
+                    const char *generic_name);
+extern inline int iioutils_get_type(unsigned *is_signed,
+                 unsigned *bytes,
+                 unsigned *bits_used,
+                 unsigned *shift,
+                 uint64_t *mask,
+                 unsigned *be,
+                 const char *device_dir,
+                 const char *name,
+                 const char *generic_name);
+extern inline void bsort_channel_array_by_index(struct iio_channel_info **ci_array,
+                     int cnt);
 
 /**
  * size_from_channelarray() - calculate the storage size of a scan
@@ -124,7 +145,7 @@ void process_scan(char *data,
 			  /*pfps printf("MASK %d %8x  ",channels[k].bits_used,val); */
 			  val = (int32_t)(val << (32 - channels[k].bits_used)) >> (32 - channels[k].bits_used);
 			  /*pfps printf("FIX %x\n",val); */
-			  printf("%s %4d %6.1f  ", channels[k].name, 
+			  printf("%s %4d %6.1f  ", channels[k].name,
 				 val, ((float)val + channels[k].offset)* channels[k].scale);
 			}
                         break;
@@ -273,7 +294,7 @@ int main(int argc, char **argv)
         printf("iio device number being used is %d\n", dev_num);
 
         asprintf(&dev_dir_name, "%siio:device%d", iio_dir, dev_num);
-	
+
 	/* enable the sensors in the device */
 	enable_sensors(dev_dir_name);
 

--- a/sensors/hinge.c
+++ b/sensors/hinge.c
@@ -1,4 +1,4 @@
-/* Rotate Lenovo Yoga (2 Pro) display and ELAN Touchscreen to match hinge of screen 
+/* Rotate Lenovo Yoga (2 Pro) display and ELAN Touchscreen to match hinge of screen
  * Copyright (c) 2014 Peter F. Patel-Schneider
  *
  * Modified from industrialio buffer test code.
@@ -19,7 +19,7 @@
  * hinge is the beginnings of a system to determine the hinge angle from the
  * two accelerometers.  It is not functional as of yet.
  *
- * WARNING:  This is not production quality code.  
+ * WARNING:  This is not production quality code.
  */
 
 #define _GNU_SOURCE
@@ -44,6 +44,28 @@
 #include "iio_utils.h"
 
 #include <X11/extensions/Xrandr.h>
+
+extern inline int build_channel_array(const char *device_dir,
+                  struct iio_channel_info **ci_array,
+                  int *counter);
+extern inline int iioutils_break_up_name(const char *full_name,
+          char **generic_name);
+extern inline int iioutils_get_param_float(float *output,
+            const char *param_name,
+            const char *device_dir,
+            const char *name,
+            const char *generic_name);
+extern inline int iioutils_get_type(unsigned *is_signed,
+           unsigned *bytes,
+           unsigned *bits_used,
+           unsigned *shift,
+           uint64_t *mask,
+           unsigned *be,
+           const char *device_dir,
+           const char *name,
+           const char *generic_name);
+extern inline void bsort_channel_array_by_index(struct iio_channel_info **ci_array,
+           int cnt);
 
 
 static int debug_level = 0;
@@ -92,7 +114,7 @@ int size_from_channelarray(struct iio_channel_info *channels, int num_channels)
  *  location offsets.
  * @num_channels:       number of channels
  * ch_name:		name of channel to get
- * ch_val:		value for the channel 
+ * ch_val:		value for the channel
  * ch_present:		whether the channel is present
  **/
 void process_scan_1(char *data, struct iio_channel_info *channels, int num_channels,
@@ -130,7 +152,7 @@ void process_scan_1(char *data, struct iio_channel_info *channels, int num_chann
 
 
 /**
- * process_scan_3() - get three integer values - see above 
+ * process_scan_3() - get three integer values - see above
  **/
 void process_scan_3(char *data, struct iio_channel_info *channels, int num_channels,
 		    char *ch_name_1, int *ch_val_1, bool *ch_present_1,
@@ -311,13 +333,13 @@ int find_orientation(int dev_num, char * dev_dir_name, char * trigger_name,
   } else orientation = accel_x > 0 ? LEFT : RIGHT;
   if (debug_level > 1) printf("Orientation %d: %5d %5d %5d\n",
 				    orientation,accel_x,accel_y,accel_z);
-  
+
   return(orientation);
 }
 
 
 /* symbolic orientation as used in xrandr */
-char * symbolic_orientation(orientation) { 
+char * symbolic_orientation(orientation) {
   char * orient;
   switch ( orientation ) {
   case FLAT : orient = "flat"; break;
@@ -423,7 +445,7 @@ int setup_device(const char * dev_name, char **dev_dir_name, char **dev_trigger_
     ret = -ENODEV;
     goto error_ret;
   }
-  
+
   if (debug_level > -1) printf("iio device number being used is %d\n", *dev_num);
   /* enable the sensors */
   asprintf(dev_dir_name, "%siio:device%d", iio_dir, *dev_num);

--- a/sensors/light.c
+++ b/sensors/light.c
@@ -30,6 +30,29 @@
 #include "libs/config.h"
 #include "libs/device_info.h"
 
+extern inline int build_channel_array(const char *device_dir,
+                  struct iio_channel_info **ci_array,
+                  int *counter);
+extern inline int iioutils_break_up_name(const char *full_name,
+				  char **generic_name);
+extern inline int iioutils_get_param_float(float *output,
+				    const char *param_name,
+				    const char *device_dir,
+				    const char *name,
+				    const char *generic_name);
+extern inline int iioutils_get_type(unsigned *is_signed,
+			     unsigned *bytes,
+			     unsigned *bits_used,
+			     unsigned *shift,
+			     uint64_t *mask,
+			     unsigned *be,
+			     const char *device_dir,
+			     const char *name,
+			     const char *generic_name);
+extern inline void bsort_channel_array_by_index(struct iio_channel_info **ci_array,
+					 int cnt);
+
+
 /**
  * process_scan() - print out the values in SI units
  * @data:               pointer to the start of the scan

--- a/sensors/orientation.c
+++ b/sensors/orientation.c
@@ -30,6 +30,28 @@
 
 #include "libs/common.h"
 
+extern inline int build_channel_array(const char *device_dir,
+                  struct iio_channel_info **ci_array,
+                  int *counter);
+extern inline int iioutils_break_up_name(const char *full_name,
+				  char **generic_name);
+extern inline int iioutils_get_param_float(float *output,
+				    const char *param_name,
+				    const char *device_dir,
+				    const char *name,
+				    const char *generic_name);
+extern inline int iioutils_get_type(unsigned *is_signed,
+			     unsigned *bytes,
+			     unsigned *bits_used,
+			     unsigned *shift,
+			     uint64_t *mask,
+			     unsigned *be,
+			     const char *device_dir,
+			     const char *name,
+			     const char *generic_name);
+extern inline void bsort_channel_array_by_index(struct iio_channel_info **ci_array,
+					 int cnt);
+
 typedef enum {
 	INVALID = -1, FLAT = 0, TOP, RIGHT, BOTTOM, LEFT
 } OrientationPositions; /* various orientations */

